spec.md

```diff
@@ -0,0 +1,954 @@
+
+- If a VM corresponding to the specified `MachineID` exists on provider the `Exists` field in the response must be set to `True` else the field is to be set to `False`.
+- If a VM exists, the Plugin is expected to return a value in the `MachineStatus` field. The possible values for the same is provided by the enum below.
+- The plugin can OPTIONALY make use of the secrets supplied in the `Secrets` map in the `GetMachineRequest` to communicate with the provider.
+- This operation MUST be idempotent.
+
+```protobuf
+message GetMachineRequest {
+    // MachineID is the id of the machine whose status is to be determined.
+    // It should uniquely identify the real machine in cloud-provider.
+    // This field is REQUIRED.
+    string MachineID = 1;
+
+    // Secrets is the map containing necessary credentials for cloud-provider to list the machines.
+    // This field is OPTIONAL.
+    map<string, bytes> Secrets = 2 [(cmi_secret) = true];
+}
+
+message GetMachineResponse {
+    // Exists tells if the machine exists on the cloud provider
+    // This field is REQUIRED.
+    bool Exists = 1;
+
+    // Status contains the possible status for machines running on the cloud
+    enum Status {
+        Unknown = 0;
+        Running = 1;
+        Stopped = 2;
+    }
+    Status MachineStatus = 2;
+}
+```
+
+##### GetMachine Errors
+
+If the plugin is unable to complete the GetMachine call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+
+| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
+|-----------|-----------|-------------|-------------------|------------|
+| 0 OK | Successful | The call was successful in getting machine details for given `MachineID` |  | N |
+| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
+| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
+| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `MachineID` and make sure that it is in the desired format and not a blank value. Exact issue to be given in `.message` | Update `MachineID` to fix issues. | N |
+| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |
+| 5 NOT_FOUND |  |  |  |  |
+| 6 ALREADY_EXISTS |  |  |  |  |
+| 7 PERMISSION_DENIED | Insufficent permissions | The requestor doesn't have enough permissions to get details for the VM and it's required dependencies | Update requestor permissions to grant the same | N |
+| 8 RESOURCE_EXHAUSTED |  |  |  |  |
+| 9 PRECONDITION_FAILED | VM is in inconsistent state | The VM is in a state that is invalid for this operation | Manual intervention might be needed to fix the state of the VM | N |
+| 10 ABORTED |  |  |  |  |
```

Comment:
Not sure about `ABORTED` for `GetMachines`. We would not return it in our implementation since it's harmless to query GMP machines in parallel. I think you should decide on one of the following 2 strategies and describe accordingly:
* `ABORTED` is valid only for operations that change the state of a machine.
* `ABORTED` is valid everywhere.